ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs

```diff
@@ -198,16 +215,126 @@ instance ShelleyBasedEra era => HasPartialLedgerConfig (ShelleyBlock era) where
   type PartialLedgerConfig (ShelleyBlock era) = ShelleyPartialLedgerConfig era
 
   -- Replace the dummy 'EpochInfo' with the real one
-  completeLedgerConfig _ epochInfo (ShelleyPartialLedgerConfig cfg _) =
+  completeLedgerConfig _ epochInfo_ (ShelleyPartialLedgerConfig cfg _) =
       cfg {
           shelleyLedgerGlobals = (shelleyLedgerGlobals cfg) {
               SL.epochInfoWithErr =
                   hoistEpochInfo
                     (runExcept . withExceptT (T.pack . show))
-                    epochInfo
+                    epochInfo_
             }
         }
 
+-- | This instance uses the invariant that the EpochInfo in a
+-- ShelleyLedgerConfig is fixed i.e. has a constant EpochSize and SlotLength.
+-- This is not true in the case of the HFC in a ShelleyPartialLedgerConfig, but
+-- that is handled correctly in the respective SerialiseNodeToClient instance
+-- for ShelleyPartialLedgerConfig.
+instance ShelleyBasedEra era
+      => SerialiseNodeToClient (ShelleyBlock era) (ShelleyLedgerConfig era) where
+  decodeNodeToClient ccfg version = do
+    enforceSize "ShelleyLedgerConfig" 3
+    partialConfig <- decodeNodeToClient
+      @_
+      @(ShelleyPartialLedgerConfig era)
+      ccfg
+      version
+    epochSize <- fromCBOR @EpochSize
+    slotLength <- decode @SlotLength
+    return $ completeLedgerConfig
+      (Proxy @(ShelleyBlock era))
+      (fixedEpochInfo epochSize slotLength)
+      partialConfig
+
+  encodeNodeToClient ccfg version ledgerConfig = mconcat [
+      encodeListLen 3
+    , encodeNodeToClient
+        @_
+        @(ShelleyPartialLedgerConfig era)
+        ccfg
+        version
+        (toPartialLedgerConfig (Proxy @(ShelleyBlock era)) ledgerConfig)
+    , toCBOR @EpochSize epochSize
+    , encode @SlotLength slotLength
+    ]
+    where
+      unwrap = either
+        (error "ShelleyLedgerConfig contains a non-fixed EpochInfo")
+        id
+      epochInfo_ = epochInfoWithErr (shelleyLedgerGlobals ledgerConfig)
+      epochSize = unwrap $ epochInfoSize epochInfo_ 0
+      RelativeTime t1 = unwrap $ epochInfoSlotToRelativeTime epochInfo_ 1
+      slotLength = mkSlotLength t1
+
+-- | This instance uses the invariant that the EpochInfo in a
+-- ShelleyPartialLedgerConfig is always just a dummy value.
+instance ShelleyBasedEra era
```

Comment:
`decode` expects 13 and `encode` expects 15 but I count 14 😆

Reply:
Wow. yes. But how could this have passed the golden tests?! Those should check that the CBOR is valid.

Reply:
Ok, So I'm looking at `Test.Util.Serialisation.Golden`'s `goldenTestCBOR`. It seems that we do some amount of checking for valid CBOR, but this is not extensive.

In the case of the golden file, we only check that we're able to decode a series of `TermToken`s (see `decodeAsFlatTerm`). We check the individual tokens are valid, but not that the sequence of tokens are valid as a whole. Things like incorrect list length are not detected (hence why this passed the golden tests)!

In the case of the actual value, we only check that the encoding doesn't throw an exception (though this is still a passed test if the golden file contains the same exception).

Note that we never actually test `decodeNodeToClient`, so the `enforceSize` call never gets a chance to check the list length either. Yes, you heard me right! Our golden tests only test the encode, not the decoder!

What I propose we do here is:
* [ ] I fix this instance with the correct list length of 14 (yes I double checked)
* [ ] In a separate PR I change goldenTestCBOR to...
  * ... use [validFlatTerm](https://hackage.haskell.org/package/cborg-0.2.5.0/docs/Codec-CBOR-FlatTerm.html#v:validFlatTerm) to check that both the "golden" decoding and the "actual" encoding are valid terms. I hope there are no existing golden tests with invalid CBOR :grimacing:.
  * ... decode the golden file using  `decodeNodeToClient` to test the decoding half of our serialization code.

Reply:
We have separate Roundtrip tests; are we just missing those for this type maybe?

Reply:
I'd say you're right, I should add round trip tests, but also it seems odd to me that we only test serializes and not deserializers in the golden tests. We really do want to test the the deserializers will succeed on the golden files.

Reply:
I have a slight preference for not testing the golden files deserialize -- leaving the Round-Trip Tests to find bugs there. ... but your idea would catch an "out-of-date" golden test file: suppose an invariant is added to the deserializer (and the generators) but we forget to update the golden files. Let's: compute is cheap here 👍

Reply:
So I think in terms of this PR, the tests are sufficient. One might say "we're golden" :laughing:.

My concern with not testing deserialization of golden files was that our deserializers aren't tested on old node-to-client versions. You're saying the golden tests are sufficient. Indeed, the golden tests run across all supported node-to-client versions. My thought was then: but what if we change our serializers *and* deserializers in a compatible way for an old node-to-clinet version? The round trip tests will pass but we're not really testing against the correct serialization. But indeed, we'd have to change the serialization code in this scenario so a gold test that only tests serialization would catch that regression. So perhaps you're right! I'm happy to leave this as is for now.

Reply:
Sounds good, thanks. Resolving.